### ISSUE TITLE
Stochastic Gradient Tree encountered ZeroDivisionError during run

### ISSUE THREAD

**Issue by robme-l** [online-ml/river#1437]
## Versions

**river version**: 0.19 (latest)
**Python version**: 3.10.10
**Operating system**: Linux

## Describe the bug

During a classification task a Stochastic Gradient Tree yielded a `ZeroDivisionError` from within the `sgt_quantizer.py` file.

## Steps/code to reproduce

```python
# Sample code to reproduce the problem
# Please do your best to provide a Minimal, Reproducible Example: https://stackoverflow.com/help/minimal-reproducible-example
```
The SGT was defined like this:
`sgt = compose.Pipeline(('model',multiclass.OneVsRestClassifier(tree.SGTClassifier())))`

Unfortunately I am not at liberty to share the data used, however it was all continuous numeric data along with a single feature which was a boolean. Training and evaluation were done with `iter_progressive_val_score` with the default settings and `yield_predictions=True`.

Like in the code examples I stored the iterator into a variable called `steps` like so: `steps=iter_progressive_val_score(...)` and then iterated over it using `for step in steps`, processing the information yielded by predictions at every step. The problem seems to occur while retrieving the next step. I was running this on a Google Colab notebook so paste the relevant part of the traceback below:

```
[<ipython-input-6-d2700249f8af>](https://localhost:8080/#) in iter_progressive_val_score(dataset, model, metric, moment, delay, step, measure_time, measure_memory, yield_predictions)
    213     """
    214 
--> 215     yield from _progressive_validation(
    216         dataset,
    217         model,

[<ipython-input-6-d2700249f8af>](https://localhost:8080/#) in _progressive_validation(dataset, model, metric, checkpoints, moment, delay, measure_time, measure_memory, yield_predictions)
     85             n_samples_learned += 1
     86             if model._supervised:
---> 87                 model.learn_one(x, y, **kwargs)
     88             else:
     89                 model.learn_one(x, **kwargs)

[/usr/local/lib/python3.10/site-packages/river/compose/pipeline.py](https://localhost:8080/#) in learn_one(self, x, y, **params)
    472             # This is usually the last step of the pipeline.
    473             elif step._supervised:
--> 474                 step.learn_one(x=x, y=y)
    475             # Here the step is not a transformer, and it's unsupervised, such as a KMeans. This
    476             # is also usually the last step of the pipeline.

[/usr/local/lib/python3.10/site-packages/river/multiclass/ovr.py](https://localhost:8080/#) in learn_one(self, x, y, **kwargs)
     86         # Train each label's associated classifier
     87         for label, model in self.classifiers.items():
---> 88             model.learn_one(x, y == label, **kwargs)
     89 
     90         return self

[/usr/local/lib/python3.10/site-packages/river/tree/stochastic_gradient_tree.py](https://localhost:8080/#) in learn_one(self, x, y, w)
    122             y_pred_raw = self.loss_func.transfer(node.prediction())
    123             grad_hess = self.loss_func.compute_derivatives(y_true_trs, y_pred_raw)
--> 124             node.update(x, grad_hess, self, w)
    125 
    126             if node.total_weight - node.last_split_attempt_at < self.grace_period:

[/usr/local/lib/python3.10/site-packages/river/tree/nodes/sgt_nodes.py](https://localhost:8080/#) in update(self, x, gh, sgt, w)
     71             else:
     72                 try:
---> 73                     self._split_stats[idx].update(x_val, gh, w)  # type: ignore
     74                 except KeyError:
     75                     # Create a new quantizer

[/usr/local/lib/python3.10/site-packages/river/tree/splitter/sgt_quantizer.py](https://localhost:8080/#) in update(self, x_val, gh, w)
    137             # Replay buffer
    138             for x_val, gh, w in self._buffer:
--> 139                 pos = math.floor((x_val - self._min) / self._radius)
    140                 if pos >= self.n_bins:
    141                     pos = self.n_bins - 1

ZeroDivisionError: float division by zero
```

Note: this error did not occur on other datasets that were similarly constructed.

**Comment by robme-l** [comment]
Following up that this occurs over a couple datasets (they all have the same characteristics and features), I will try to analyze the datasets and see why this happens. All the other RiverML models work fine by the way.

**Comment by smastelini** [comment]
Hi @robme-l , I suggest you pass the boolean feature identifier to the `nominal_attributes` parameter. If that doesn't work, the problem should be related to low variability in your variables. 

SGTs quantize the data using either a fixed amount of points to create intervals or the standard deviation of the features (by default). If in any of the cases the variance is zero you will get this error.

**Comment by robme-l** [comment]
@smastelini I got some time this week to inspect the datasets and you are right, the variance (or standard deviation in general) is extremely low, especially when compared to the other datasets. However none of their variance's were 0 which was odd. However I assume that low-variability was the issue. Either way, I opted not to use SGT in further trials due to lower performance on the given task compared to even the Dummy variable so leaving it at that.
